aws_ami: gp3 volume scalable throughtput

ewbankkit opened the pull request:
<!--- See what makes a good Pull Request at : https://github.com/hashicorp/terraform-provider-aws/blob/master/docs/CONTRIBUTING.md --->

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

<!--- Thank you for keeping this note for the community --->

<!--- If your PR fully resolves and should automatically close the linked issue, use Closes. Otherwise, use Relates --->
Relates #16514.

Release note for [CHANGELOG](https://github.com/hashicorp/terraform-provider-aws/blob/master/CHANGELOG.md):
<!--
If change is not user facing, just write "NONE" in the release-note block below.
-->

```release-note
resource/aws_ami: Support `volume_type` value of `gp3`. Add `throughput` attribute to `ebs_block_device` configuration.
data-source/aws_ami: Add `throughput` value to `block_device_mappings.ebs` mapping.
resource/aws_ami_copy: Add `throughput` attribute to `ebs_block_device` configuration.
resource/aws_ami_from_instance: Add `throughput` attribute to `ebs_block_device` configuration.
```

Output from acceptance testing:

<!--
Replace TestAccXXX with a pattern that matches the tests affected by this PR.

For more information on the `-run` flag, see the `go test` documentation at https://tip.golang.org/cmd/go/#hdr-Testing_flags.
-->
```console
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSAMI_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSAMI_ -timeout 120m
=== RUN   TestAccAWSAMI_basic
=== PAUSE TestAccAWSAMI_basic
=== RUN   TestAccAWSAMI_description
=== PAUSE TestAccAWSAMI_description
=== RUN   TestAccAWSAMI_disappears
=== PAUSE TestAccAWSAMI_disappears
=== RUN   TestAccAWSAMI_EphemeralBlockDevices
=== PAUSE TestAccAWSAMI_EphemeralBlockDevices
=== RUN   TestAccAWSAMI_Gp3BlockDevice
=== PAUSE TestAccAWSAMI_Gp3BlockDevice
=== RUN   TestAccAWSAMI_tags
=== PAUSE TestAccAWSAMI_tags
=== CONT  TestAccAWSAMI_description
=== CONT  TestAccAWSAMI_basic
=== CONT  TestAccAWSAMI_Gp3BlockDevice
=== CONT  TestAccAWSAMI_tags
=== CONT  TestAccAWSAMI_EphemeralBlockDevices
=== CONT  TestAccAWSAMI_disappears
--- PASS: TestAccAWSAMI_description (57.76s)
--- PASS: TestAccAWSAMI_disappears (61.16s)
--- PASS: TestAccAWSAMI_EphemeralBlockDevices (64.17s)
--- PASS: TestAccAWSAMI_basic (64.66s)
--- PASS: TestAccAWSAMI_Gp3BlockDevice (65.80s)
--- PASS: TestAccAWSAMI_tags (86.83s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	86.928s
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSAmiDataSource_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSAmiDataSource_ -timeout 120m
=== RUN   TestAccAWSAmiDataSource_natInstance
=== PAUSE TestAccAWSAmiDataSource_natInstance
=== RUN   TestAccAWSAmiDataSource_windowsInstance
=== PAUSE TestAccAWSAmiDataSource_windowsInstance
=== RUN   TestAccAWSAmiDataSource_instanceStore
=== PAUSE TestAccAWSAmiDataSource_instanceStore
=== RUN   TestAccAWSAmiDataSource_localNameFilter
=== PAUSE TestAccAWSAmiDataSource_localNameFilter
=== RUN   TestAccAWSAmiDataSource_Gp3BlockDevice
=== PAUSE TestAccAWSAmiDataSource_Gp3BlockDevice
=== CONT  TestAccAWSAmiDataSource_natInstance
=== CONT  TestAccAWSAmiDataSource_localNameFilter
=== CONT  TestAccAWSAmiDataSource_Gp3BlockDevice
=== CONT  TestAccAWSAmiDataSource_instanceStore
=== CONT  TestAccAWSAmiDataSource_windowsInstance
--- PASS: TestAccAWSAmiDataSource_natInstance (16.66s)
--- PASS: TestAccAWSAmiDataSource_instanceStore (16.93s)
--- PASS: TestAccAWSAmiDataSource_windowsInstance (17.75s)
--- PASS: TestAccAWSAmiDataSource_localNameFilter (18.27s)
--- PASS: TestAccAWSAmiDataSource_Gp3BlockDevice (56.17s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	56.257s
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSAMICopy_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSAMICopy_ -timeout 120m
=== RUN   TestAccAWSAMICopy_basic
=== PAUSE TestAccAWSAMICopy_basic
=== RUN   TestAccAWSAMICopy_Description
=== PAUSE TestAccAWSAMICopy_Description
=== RUN   TestAccAWSAMICopy_EnaSupport
=== PAUSE TestAccAWSAMICopy_EnaSupport
=== RUN   TestAccAWSAMICopy_tags
=== PAUSE TestAccAWSAMICopy_tags
=== CONT  TestAccAWSAMICopy_basic
=== CONT  TestAccAWSAMICopy_tags
=== CONT  TestAccAWSAMICopy_EnaSupport
=== CONT  TestAccAWSAMICopy_Description
--- PASS: TestAccAWSAMICopy_EnaSupport (385.38s)
--- PASS: TestAccAWSAMICopy_basic (386.46s)
--- PASS: TestAccAWSAMICopy_Description (397.85s)
--- PASS: TestAccAWSAMICopy_tags (411.06s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	411.151s
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSAMIFromInstance_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSAMIFromInstance_ -timeout 120m
=== RUN   TestAccAWSAMIFromInstance_basic
=== PAUSE TestAccAWSAMIFromInstance_basic
=== RUN   TestAccAWSAMIFromInstance_tags
=== PAUSE TestAccAWSAMIFromInstance_tags
=== CONT  TestAccAWSAMIFromInstance_basic
=== CONT  TestAccAWSAMIFromInstance_tags
--- PASS: TestAccAWSAMIFromInstance_basic (243.32s)
--- PASS: TestAccAWSAMIFromInstance_tags (302.97s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	303.062s
```

bflad reviewed (changes requested):
Sorry this is not easy to know right now, but the `aws_ami_copy` and `aws_ami_from_instance` resources also require updates when adjusting this:

```
=== CONT  TestAccAWSAMIFromInstance_basic
------- Stderr: -------
panic: Invalid address to set: []string{"ebs_block_device", "0", "throughput"}
goroutine 447 [running]:
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*ResourceData).Set(0xc001c22c00, 0x77540f5, 0x10, 0x60b8ce0, 0xc001c73200, 0x1, 0x0)
  /opt/teamcity-agent/work/2e10e023da0c7520/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.3.0/helper/schema/resource_data.go:230 +0x36f
github.com/terraform-providers/terraform-provider-aws/aws.resourceAwsAmiRead(0xc001c22c00, 0x69f6660, 0xc000694000, 0xc0025a7b50, 0xc00064ab00)
  /opt/teamcity-agent/work/2e10e023da0c7520/src/github.com/terraform-providers/terraform-provider-aws/aws/resource_aws_ami.go:366 +0xd47
github.com/terraform-providers/terraform-provider-aws/aws.resourceAwsAmiFromInstanceCreate(0xc001c22c00, 0x69f6660, 0xc000694000, 0x3, 0xffffffffffffffff)
  /opt/teamcity-agent/work/2e10e023da0c7520/src/github.com/terraform-providers/terraform-provider-aws/aws/resource_aws_ami_from_instance.go:215 +0x425
```

```
=== CONT  TestAccAWSAMICopy_basic
------- Stderr: -------
panic: Invalid address to set: []string{"ebs_block_device", "0", "throughput"}
goroutine 627 [running]:
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*ResourceData).Set(0xc00231f980, 0x77540f5, 0x10, 0x60b8ce0, 0xc001cbcd40, 0x1, 0x0)
  /opt/teamcity-agent/work/2e10e023da0c7520/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.3.0/helper/schema/resource_data.go:230 +0x36f
github.com/terraform-providers/terraform-provider-aws/aws.resourceAwsAmiRead(0xc00231f980, 0x69f6660, 0xc001e48680, 0xc0031af528, 0xc002d80e00)
  /opt/teamcity-agent/work/2e10e023da0c7520/src/github.com/terraform-providers/terraform-provider-aws/aws/resource_aws_ami.go:366 +0xd47
github.com/terraform-providers/terraform-provider-aws/aws.resourceAwsAmiCopyCreate(0xc00231f980, 0x69f6660, 0xc001e48680, 0x2, 0xffffffffffffffff)
  /opt/teamcity-agent/work/2e10e023da0c7520/src/github.com/terraform-providers/terraform-provider-aws/aws/resource_aws_ami_copy.go:233 +0x4e7
```

Hopefully early next year we will have some tooling to help with this, even for running acceptance tests locally.

ewbankkit commented:
* Modified the new `expand` and `flatten` functions to follow the [Data Handling and Conversion guidelines](https://github.com/hashicorp/terraform-provider-aws/blob/master/docs/contributing/data-handling-and-conversion.md)
* Added the `enabled` and `snapshot_id` conflict checking

Acceptance test output:

```console
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSAMI_\|TestAccAWSAmiDataSource_\|TestAccAWSAMICopy_\|TestAccAWSAMIFromInstance_' ACCTEST_PARALLELISM=4
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 4 -run=TestAccAWSAMI_\|TestAccAWSAmiDataSource_\|TestAccAWSAMICopy_\|TestAccAWSAMIFromInstance_ -timeout 120m
=== RUN   TestAccAWSAmiDataSource_natInstance
=== PAUSE TestAccAWSAmiDataSource_natInstance
=== RUN   TestAccAWSAmiDataSource_windowsInstance
=== PAUSE TestAccAWSAmiDataSource_windowsInstance
=== RUN   TestAccAWSAmiDataSource_instanceStore
=== PAUSE TestAccAWSAmiDataSource_instanceStore
=== RUN   TestAccAWSAmiDataSource_localNameFilter
=== PAUSE TestAccAWSAmiDataSource_localNameFilter
=== RUN   TestAccAWSAmiDataSource_Gp3BlockDevice
=== PAUSE TestAccAWSAmiDataSource_Gp3BlockDevice
=== RUN   TestAccAWSAMICopy_basic
=== PAUSE TestAccAWSAMICopy_basic
=== RUN   TestAccAWSAMICopy_Description
=== PAUSE TestAccAWSAMICopy_Description
=== RUN   TestAccAWSAMICopy_EnaSupport
=== PAUSE TestAccAWSAMICopy_EnaSupport
=== RUN   TestAccAWSAMICopy_tags
=== PAUSE TestAccAWSAMICopy_tags
=== RUN   TestAccAWSAMIFromInstance_basic
=== PAUSE TestAccAWSAMIFromInstance_basic
=== RUN   TestAccAWSAMIFromInstance_tags
=== PAUSE TestAccAWSAMIFromInstance_tags
=== RUN   TestAccAWSAMI_basic
=== PAUSE TestAccAWSAMI_basic
=== RUN   TestAccAWSAMI_description
=== PAUSE TestAccAWSAMI_description
=== RUN   TestAccAWSAMI_disappears
=== PAUSE TestAccAWSAMI_disappears
=== RUN   TestAccAWSAMI_EphemeralBlockDevices
=== PAUSE TestAccAWSAMI_EphemeralBlockDevices
=== RUN   TestAccAWSAMI_Gp3BlockDevice
=== PAUSE TestAccAWSAMI_Gp3BlockDevice
=== RUN   TestAccAWSAMI_tags
=== PAUSE TestAccAWSAMI_tags
=== CONT  TestAccAWSAmiDataSource_natInstance
=== CONT  TestAccAWSAMI_tags
=== CONT  TestAccAWSAMI_Gp3BlockDevice
=== CONT  TestAccAWSAMI_EphemeralBlockDevices
--- PASS: TestAccAWSAmiDataSource_natInstance (14.28s)
=== CONT  TestAccAWSAMI_disappears
--- PASS: TestAccAWSAMI_disappears (39.57s)
=== CONT  TestAccAWSAMI_description
--- PASS: TestAccAWSAMI_Gp3BlockDevice (61.35s)
=== CONT  TestAccAWSAMI_basic
--- PASS: TestAccAWSAMI_EphemeralBlockDevices (61.73s)
=== CONT  TestAccAWSAMIFromInstance_tags
--- PASS: TestAccAWSAMI_tags (86.70s)
=== CONT  TestAccAWSAMICopy_basic
--- PASS: TestAccAWSAMI_basic (56.36s)
=== CONT  TestAccAWSAMIFromInstance_basic
--- PASS: TestAccAWSAMI_description (69.73s)
=== CONT  TestAccAWSAMICopy_tags
=== CONT  TestAccAWSAMICopy_EnaSupport
--- PASS: TestAccAWSAMIFromInstance_basic (212.82s)
--- PASS: TestAccAWSAMIFromInstance_tags (285.49s)
=== CONT  TestAccAWSAMICopy_Description
=== CONT  TestAccAWSAmiDataSource_localNameFilter
--- PASS: TestAccAWSAMICopy_basic (370.81s)
--- PASS: TestAccAWSAmiDataSource_localNameFilter (13.49s)
=== CONT  TestAccAWSAmiDataSource_Gp3BlockDevice
--- PASS: TestAccAWSAmiDataSource_Gp3BlockDevice (41.36s)
=== CONT  TestAccAWSAmiDataSource_instanceStore
--- PASS: TestAccAWSAMICopy_tags (398.78s)
=== CONT  TestAccAWSAmiDataSource_windowsInstance
--- PASS: TestAccAWSAmiDataSource_instanceStore (11.85s)
--- PASS: TestAccAWSAmiDataSource_windowsInstance (12.87s)
--- PASS: TestAccAWSAMICopy_Description (384.37s)
--- PASS: TestAccAWSAMICopy_EnaSupport (438.88s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	769.512s
```

bflad reviewed (approved):
Looks good to me 🚀  (Won't merge immediately as we are in a temporary code freeze until the aws_instance regression and https://github.com/hashicorp/terraform-provider-aws/pull/16215 are sorted)

Output from acceptance testing in AWS Commercial:

```
--- PASS: TestAccAWSAMI_basic (64.21s)
--- PASS: TestAccAWSAMI_description (76.85s)
--- PASS: TestAccAWSAMI_disappears (60.98s)
--- PASS: TestAccAWSAMI_EphemeralBlockDevices (64.91s)
--- PASS: TestAccAWSAMI_Gp3BlockDevice (48.76s)
--- PASS: TestAccAWSAMI_tags (73.71s)

--- PASS: TestAccAWSAMICopy_basic (378.01s)
--- PASS: TestAccAWSAMICopy_Description (391.97s)
--- PASS: TestAccAWSAMICopy_EnaSupport (377.97s)
--- PASS: TestAccAWSAMICopy_tags (404.08s)

--- PASS: TestAccAWSAmiDataSource_Gp3BlockDevice (48.44s)
--- PASS: TestAccAWSAmiDataSource_instanceStore (27.12s)
--- PASS: TestAccAWSAmiDataSource_localNameFilter (30.76s)
--- PASS: TestAccAWSAmiDataSource_natInstance (26.80s)
--- PASS: TestAccAWSAmiDataSource_windowsInstance (28.30s)

--- PASS: TestAccAWSAMIFromInstance_basic (444.76s)
--- PASS: TestAccAWSAMIFromInstance_tags (294.34s)

--- PASS: TestAccAWSAMILaunchPermission_basic (344.81s)
--- PASS: TestAccAWSAMILaunchPermission_Disappears_AMI (359.20s)
--- PASS: TestAccAWSAMILaunchPermission_Disappears_LaunchPermission (336.46s)
--- PASS: TestAccAWSAMILaunchPermission_Disappears_LaunchPermission_Public (348.74s)
```

Output from acceptance testing in AWS GovCloud (US):

```
--- PASS: TestAccAWSAMI_basic (68.61s)
--- PASS: TestAccAWSAMI_description (76.31s)
--- PASS: TestAccAWSAMI_disappears (65.32s)
--- PASS: TestAccAWSAMI_EphemeralBlockDevices (69.63s)
--- PASS: TestAccAWSAMI_Gp3BlockDevice (67.43s)
--- PASS: TestAccAWSAMI_tags (99.68s)

--- PASS: TestAccAWSAMICopy_basic (391.97s)
--- PASS: TestAccAWSAMICopy_Description (409.20s)
--- PASS: TestAccAWSAMICopy_EnaSupport (392.72s)
--- PASS: TestAccAWSAMICopy_tags (413.77s)

--- PASS: TestAccAWSAmiDataSource_Gp3BlockDevice (67.39s)
--- PASS: TestAccAWSAmiDataSource_instanceStore (27.26s)
--- PASS: TestAccAWSAmiDataSource_localNameFilter (30.92s)
--- PASS: TestAccAWSAmiDataSource_natInstance (27.50s)
--- PASS: TestAccAWSAmiDataSource_windowsInstance (27.77s)

--- PASS: TestAccAWSAMIFromInstance_basic (255.31s)
--- PASS: TestAccAWSAMIFromInstance_tags (359.49s)

--- PASS: TestAccAWSAMILaunchPermission_basic (356.36s)
--- PASS: TestAccAWSAMILaunchPermission_Disappears_AMI (371.79s)
--- PASS: TestAccAWSAMILaunchPermission_Disappears_LaunchPermission (354.32s)
--- PASS: TestAccAWSAMILaunchPermission_Disappears_LaunchPermission_Public (353.67s)
```
